dummyserver/hypercornserver.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import annotations
+
+import concurrent.futures
+import contextlib
+import functools
+import sys
+import threading
+from typing import Generator
+
+import hypercorn
+import hypercorn.trio
+import trio
+from quart_trio import QuartTrio
+
+
+# https://github.com/pgjones/hypercorn/blob/19dfb96411575a6a647cdea63fa581b48ebb9180/src/hypercorn/utils.py#L172-L178
+async def graceful_shutdown(shutdown_event: threading.Event) -> None:
+    while True:
+        if shutdown_event.is_set():
+            return
+        await trio.sleep(0.1)
+
+
+async def _start_server(
+    config: hypercorn.Config,
+    app: QuartTrio,
+    ready_event: threading.Event,
+    shutdown_event: threading.Event,
+) -> None:
+    async with trio.open_nursery() as nursery:
+        await nursery.start(
+            functools.partial(
+                hypercorn.trio.serve,
+                app,
+                config,
+                shutdown_trigger=functools.partial(graceful_shutdown, shutdown_event),
+            )
+        )
+        ready_event.set()
+
+
+@contextlib.contextmanager
+def run_hypercorn_in_thread(
+    config: hypercorn.Config, app: QuartTrio
+) -> Generator[None, None, None]:
+    ready_event = threading.Event()
+    shutdown_event = threading.Event()
+
+    with concurrent.futures.ThreadPoolExecutor(
+        1, thread_name_prefix="hypercorn dummyserver"
+    ) as executor:
+        future = executor.submit(
+            trio.run,
+            _start_server,  # type: ignore[arg-type]
+            config,
+            app,
+            ready_event,
+            shutdown_event,
+        )
+        ready_event.wait(5)
```

Comment:
Should we error out if the `ready_event` doesn't trigger in the timeout time?

Reply:
Thanks, addressed in https://github.com/urllib3/urllib3/pull/3190/commits/956aa8fbd4a9d7a9df740f0ff6cc647f43e8c653